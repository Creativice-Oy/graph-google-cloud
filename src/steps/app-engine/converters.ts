import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { appengine_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_APP_ENGINE_APPLICATION,
  ENTITY_CLASS_APP_ENGINE_INSTANCE,
  ENTITY_CLASS_APP_ENGINE_SERVICE,
  ENTITY_CLASS_APP_ENGINE_VERSION,
  ENTITY_TYPE_APP_ENGINE_APPLICATION,
  ENTITY_TYPE_APP_ENGINE_INSTANCE,
  ENTITY_TYPE_APP_ENGINE_SERVICE,
  ENTITY_TYPE_APP_ENGINE_VERSION,
} from './constants';

export function createAppEngineApplicationEntity(
  data: appengine_v1.Schema$Application,
  projectId: string,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_APPLICATION,
        _type: ENTITY_TYPE_APP_ENGINE_APPLICATION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        location: data.locationId,
        servingStatus: data.servingStatus,
        defaultHostname: data.defaultHostname,
        defaultBucket: data.defaultBucket,
        splitHealthChecks: data.featureSettings?.splitHealthChecks,
        useContainerOptimizedOs: data.featureSettings?.useContainerOptimizedOs,
        webLink: getGoogleCloudConsoleWebLink(
          `/appengine?project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppEngineServiceEntity(
  data: appengine_v1.Schema$Service,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_SERVICE,
        _type: ENTITY_TYPE_APP_ENGINE_SERVICE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
      },
    },
  });
}

export function createAppEngineVersionEntity(
  data: appengine_v1.Schema$Version,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_VERSION,
        _type: ENTITY_TYPE_APP_ENGINE_VERSION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        versionUrl: data.versionUrl,
        category: ['application'],
        createdOn: parseTimePropertyValue(data.createTime),
      },
    },
  });
}

export function createAppEngineInstanceEntity(
  data: appengine_v1.Schema$Instance,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_INSTANCE,
        _type: ENTITY_TYPE_APP_ENGINE_INSTANCE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        availability: data.availability,
        requests: data.requests,
        createdOn: parseTimePropertyValue(data.startTime),
      },
    },
  });
}
