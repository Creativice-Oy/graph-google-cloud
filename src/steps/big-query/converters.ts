import { bigquery_v2 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  BIG_QUERY_DATASET_ENTITY_CLASS,
  BIG_QUERY_DATASET_ENTITY_TYPE,
  BIG_QUERY_TABLE_ENTITY_CLASS,
  BIG_QUERY_TABLE_ENTITY_TYPE,
} from './constants';
import { isMemberPublic } from '../../utils/iam';
import { BigQueryTable } from './client';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';

interface DatasetAccess {
  domain?: string;
  groupByEmail?: string;
  iamMember?: string;
  role?: string;
  specialGroup?: string;
  userByEmail?: string;
}

// 7.1 Ensure that BigQuery datasets are not anonymously or publicly accessible (Scored)
function isBigQueryDatasetPublicAccess(accessList: DatasetAccess[]): boolean {
  for (const access of accessList) {
    const value =
      access.groupByEmail ||
      access.userByEmail ||
      access.iamMember ||
      access.specialGroup ||
      '';
    if (isMemberPublic(value)) {
      return true;
    }
  }

  return false;
}

export function createBigQueryDatasetEntity(data: bigquery_v2.Schema$Dataset) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_DATASET_ENTITY_TYPE,
        _class: BIG_QUERY_DATASET_ENTITY_CLASS,
        name: data.datasetReference?.datasetId,
        description: data.description,
        public: isBigQueryDatasetPublicAccess(data.access || []),
        kmsKeyName: data.defaultEncryptionConfiguration?.kmsKeyName,
        location: data.location,
        encrypted: true,
        classification: null,
        etag: data.etag,
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        modifiedOn: data.lastModifiedTime
          ? parseInt(data.lastModifiedTime, 10)
          : undefined,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${data.datasetReference?.projectId}&d=${data.datasetReference?.datasetId}&p=${data.datasetReference?.projectId}&page=dataset`,
        ),
      },
    },
  });
}

export function createBigQueryTableEntity({
  data,
  projectId,
  isPublic,
}: {
  data: BigQueryTable;
  projectId: string;
  isPublic: boolean;
}) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_TABLE_ENTITY_TYPE,
        _class: BIG_QUERY_TABLE_ENTITY_CLASS,
        name: data.tableReference?.tableId,
        public: isPublic,
        type: data.type,
        friendlyName: data.friendlyName,
        classification: null,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${projectId}&d=${data.tableReference?.datasetId}&p=${projectId}&t=${data.tableReference?.tableId}&page=table`,
        ),
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        expirationTime: data.expirationTime
          ? parseInt(data.expirationTime, 10)
          : undefined,
      },
    },
  });
}
