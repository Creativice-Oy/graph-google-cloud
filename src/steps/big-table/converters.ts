// import { Entity } from '@jupiterone/integration-sdk-core';
import { bigtableadmin_v2 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_BIG_TABLE_APP_PROFILE,
  ENTITY_CLASS_BIG_TABLE_BACKUP,
  ENTITY_CLASS_BIG_TABLE_CLUSTER,
  ENTITY_CLASS_BIG_TABLE_INSTANCE,
  ENTITY_CLASS_BIG_TABLE_LOCATION,
  ENTITY_CLASS_BIG_TABLE_OPERATION,
  ENTITY_CLASS_BIG_TABLE_TABLE,
  ENTITY_TYPE_BIG_TABLE_APP_PROFILE,
  ENTITY_TYPE_BIG_TABLE_BACKUP,
  ENTITY_TYPE_BIG_TABLE_CLUSTER,
  ENTITY_TYPE_BIG_TABLE_INSTANCE,
  ENTITY_TYPE_BIG_TABLE_LOCATION,
  ENTITY_TYPE_BIG_TABLE_OPERATION,
  ENTITY_TYPE_BIG_TABLE_TABLE,
} from './constants';

export function getOperationKey(operation: bigtableadmin_v2.Schema$Operation) {
  return `bigtable_operation:${operation.name}`;
}

export function getInstanceKey(instance: bigtableadmin_v2.Schema$Instance) {
  return `bigtable_instance:${instance.name}`;
}

export function getAppProfileKey(
  appProfile: bigtableadmin_v2.Schema$AppProfile,
) {
  return `bigtable_appProfile:${appProfile.name}`;
}

export function getClusterKey(cluster: bigtableadmin_v2.Schema$Cluster) {
  return `bigtable_cluster:${cluster.name}`;
}

export function getBackupKey(backup: bigtableadmin_v2.Schema$Backup) {
  return `bigtable_backup:${backup.name}`;
}

export function getTableKey(table: bigtableadmin_v2.Schema$Table) {
  return `bigtable_table:${table.name}`;
}

export function getLocationKey(location: bigtableadmin_v2.Schema$Location) {
  return `bigtable_location:${location.locationId}`;
}

export function createOperationEntity({
  operation,
}: {
  operation: bigtableadmin_v2.Schema$Operation;
}) {
  return createGoogleCloudIntegrationEntity(operation, {
    entityData: {
      source: operation,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_OPERATION,
        _type: ENTITY_TYPE_BIG_TABLE_OPERATION,
        _key: getOperationKey(operation),
        name: operation.name,
        done: operation.done,
      },
    },
  });
}

export function createInstanceEntity({
  instance,
  projectId,
}: {
  instance: bigtableadmin_v2.Schema$Instance;
  projectId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(instance, {
    entityData: {
      source: instance,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_INSTANCE,
        _type: ENTITY_TYPE_BIG_TABLE_INSTANCE,
        _key: getInstanceKey(instance),
        name: instance.name,
        displayName: instance.displayName!,
        state: instance.state,
        type: instance.type,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instance.name}/overview?project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppProfileEntity({
  appProfile,
  projectId,
  instanceId,
}: {
  appProfile: bigtableadmin_v2.Schema$AppProfile;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(appProfile, {
    entityData: {
      source: appProfile,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_APP_PROFILE,
        _type: ENTITY_TYPE_BIG_TABLE_APP_PROFILE,
        _key: getAppProfileKey(appProfile),
        name: appProfile.name,
        instanceId,
        etag: appProfile.etag,
        description: appProfile.description,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/app-profiles/${appProfile.name}?project=${projectId}`,
        ),
      },
    },
  });
}

export function createClusterEntity({
  cluster,
  projectId,
  instanceId,
}: {
  cluster: bigtableadmin_v2.Schema$Cluster;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(cluster, {
    entityData: {
      source: cluster,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_CLUSTER,
        _type: ENTITY_TYPE_BIG_TABLE_CLUSTER,
        _key: getClusterKey(cluster),
        name: cluster.name,
        instanceId,
        state: cluster.state,
        serveNodes: cluster.serveNodes,
        defaultStorageType: cluster.defaultStorageType,
        location: cluster.location,
        kmsKeyName: cluster.encryptionConfig?.kmsKeyName,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/overview?project=${projectId}`,
        ),
      },
    },
  });
}

export function createBackupEntity({
  backup,
  projectId,
  instanceId,
  clusterId,
}: {
  backup: bigtableadmin_v2.Schema$Backup;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
  clusterId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(backup, {
    entityData: {
      source: backup,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_BACKUP,
        _type: ENTITY_TYPE_BIG_TABLE_BACKUP,
        _key: getBackupKey(backup),
        name: backup.name,
        instanceId,
        clusterId,
        sourceTable: backup.sourceTable,
        expireTime: backup.expireTime,
        startTime: backup.startTime,
        endTime: backup.endTime,
        sizeBytes: backup.sizeBytes,
        state: backup.state,
        encryptionType: backup.encryptionInfo?.encryptionType,
        kmsKeyVersion: backup.encryptionInfo?.kmsKeyVersion,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/backups?project=${projectId}`,
        ),
      },
    },
  });
}

export function createTableEntity({
  table,
  projectId,
  instanceId,
}: {
  table: bigtableadmin_v2.Schema$Table;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(table, {
    entityData: {
      source: table,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_TABLE,
        _type: ENTITY_TYPE_BIG_TABLE_TABLE,
        _key: getTableKey(table),
        name: table.name,
        classification: 'unclassified', // not sure what to put here. this is a required field for the class DataCollection
        instanceId,
        granularity: table.granularity || undefined, // these value are there in the type definition in the docs, but are not being sent by the API. we're only getting the name
        sourceType: table.restoreInfo?.sourceType || undefined,
        backup: table.restoreInfo?.backupInfo?.backup || undefined,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/tables?project=${projectId}`,
        ),
      },
    },
  });
}

export function createLocationEntity({
  location,
}: {
  location: bigtableadmin_v2.Schema$Location;
}) {
  return createGoogleCloudIntegrationEntity(location, {
    entityData: {
      source: location,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_LOCATION,
        _type: ENTITY_TYPE_BIG_TABLE_LOCATION,
        _key: getLocationKey(location),
        name: location.name,
        locationId: location.locationId,
        displayName: location.displayName || undefined,
      },
    },
  });
}
