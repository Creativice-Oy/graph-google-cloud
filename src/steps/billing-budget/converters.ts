import { billingbudgets_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  ENTITY_CLASS_BILLING_BUDGET,
  ENTITY_TYPE_BILLING_BUDGET,
} from './constants';

export function createBillingBudgetEntity(
  data: billingbudgets_v1.Schema$GoogleCloudBillingBudgetsV1Budget,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_BILLING_BUDGET,
        _type: ENTITY_TYPE_BILLING_BUDGET,
        _key: data.name as string,
        name: data.name,
        displayName: data.displayName as string,
      },
    },
  });
}
