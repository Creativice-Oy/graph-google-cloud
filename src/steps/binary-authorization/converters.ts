import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { binaryauthorization_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  BINARY_AUTHORIZATION_POLICY_ENTITY_CLASS,
  BINARY_AUTHORIZATION_POLICY_ENTITY_TYPE,
} from './constants';

export function createBinaryAuthorizationPolicyEntity(
  data: binaryauthorization_v1.Schema$Policy,
  projectId: string,
) {
  const admissionWhitelistPatternsSrc: binaryauthorization_v1.Schema$AdmissionWhitelistPattern[] =
    data.admissionWhitelistPatterns || [];

  const admissionWhitelistPatterns: string[] = admissionWhitelistPatternsSrc
    .map(
      (pattern: binaryauthorization_v1.Schema$AdmissionWhitelistPattern) =>
        pattern.namePattern || '',
    )
    .filter((value) => value);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: BINARY_AUTHORIZATION_POLICY_ENTITY_TYPE,
        _class: BINARY_AUTHORIZATION_POLICY_ENTITY_CLASS,
        name: data.name,
        // 6.1.4 Minimize Container Registries to only those approved (Not Scored)
        admissionWhitelistPatterns: admissionWhitelistPatterns,
        evaluationMode: data.defaultAdmissionRule?.evaluationMode,
        globalPolicyEvaluationMode: data.globalPolicyEvaluationMode,
        title: 'Binary Authorization Policy',
        summary:
          'Binary Authorization implements a policy model, where a policy is a set of rules that governs the deployment of container images to a GKE cluster. Rules in a policy provide specific criteria that an image must satisfy before it can be deployed.',
        content: '',
        webLink: getGoogleCloudConsoleWebLink(
          `/security/binary-authorization/policy?project=${projectId}&folder=&organizationId=`,
        ),
      },
    },
  });
}
