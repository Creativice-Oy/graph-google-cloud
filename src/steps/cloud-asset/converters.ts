import { cloudasset_v1 } from 'googleapis';
import { snakeCase } from 'lodash';
import { hashArray } from '../../utils/crypto';

import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { bindingEntities } from './constants';

export function buildIamBindingEntityKey({
  binding,
  projectName,
  resource,
}: {
  binding: cloudasset_v1.Schema$Binding;
  projectName: string | undefined | null;
  resource: string | undefined | null;
}) {
  // There are no unique values returned from this result so use everything
  // available to make the binding key unique.
  const keyBuilders = ['binding'];

  if (projectName) keyBuilders.push('project:' + projectName);
  if (resource) keyBuilders.push('resource:' + resource);
  if (binding.role) keyBuilders.push('role:' + binding.role);
  if (binding.members)
    keyBuilders.push('members:' + hashArray(binding.members));

  return keyBuilders.join('|');
}

export function createIamBindingEntity({
  _key,
  projectId,
  binding,
  resource,
}: {
  _key: string;
  projectId?: string;
  binding: cloudasset_v1.Schema$Binding;
  resource: string | undefined | null;
}) {
  const namePrefix = 'Role Binding for Resource: ';

  return createGoogleCloudIntegrationEntity(binding, {
    entityData: {
      source: binding,
      assign: {
        _class: bindingEntities.BINDINGS._class,
        _type: bindingEntities.BINDINGS._type,
        _key,
        name: snakeCase(namePrefix) + resource,
        displayName: namePrefix + resource,
        resource,
        role: binding.role,
        members: binding.members,
        projectId,
        'condition.title': binding.condition?.title,
        'condition.description': binding.condition?.description,
        'condition.expression': binding.condition?.expression,
      },
    },
  });
}
