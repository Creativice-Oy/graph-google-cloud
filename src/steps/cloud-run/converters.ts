import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { run_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_CLOUD_RUN_SERVICE,
  ENTITY_TYPE_CLOUD_RUN_SERVICE,
  ENTITY_TYPE_CLOUD_RUN_ROUTE,
  ENTITY_CLASS_CLOUD_RUN_ROUTE,
  ENTITY_CLASS_CLOUD_RUN_CONFIGURATION,
  ENTITY_TYPE_CLOUD_RUN_CONFIGURATION,
} from './constants';

export function createCloudRunServiceEntity(
  data: run_v1.Schema$Service,
  projectId: string,
) {
  // Build webLink
  let webLink = '';
  if (
    data.metadata?.labels &&
    data.metadata?.labels['cloud.googleapis.com/location']
  ) {
    webLink = `/run/detail/${data.metadata?.labels['cloud.googleapis.com/location']}/${data.metadata.name}`;
  } else {
    // Worst case we lead the user to "view all services" page
    webLink = `/run?project=${projectId}`;
  }

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_CLOUD_RUN_SERVICE,
        _type: ENTITY_TYPE_CLOUD_RUN_SERVICE,
        _key: data.metadata?.selfLink as string,
        name: data.metadata?.name,
        displayName: data.metadata?.name as string,
        category: ['infrastructure'],
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        webLink: getGoogleCloudConsoleWebLink(webLink),
      },
    },
  });
}

export function createCloudRunRouteEntity(data: run_v1.Schema$Route) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_CLOUD_RUN_ROUTE,
        _type: ENTITY_TYPE_CLOUD_RUN_ROUTE,
        _key: data.metadata?.selfLink as string,
        name: data.metadata?.name,
        displayName: data.metadata?.name as string,
        url: data.status?.url,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}

export function createCloudRunConfigurationEntity(
  data: run_v1.Schema$Configuration,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_CLOUD_RUN_CONFIGURATION,
        _type: ENTITY_TYPE_CLOUD_RUN_CONFIGURATION,
        _key: data.metadata?.selfLink as string,
        name: data.metadata?.name,
        displayName: data.metadata?.name as string,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
      },
    },
  });
}
