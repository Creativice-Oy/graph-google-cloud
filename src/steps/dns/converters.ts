import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import { dns_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  DNS_MANAGED_ZONE_ENTITY_CLASS,
  DNS_MANAGED_ZONE_ENTITY_TYPE,
} from './constants';

export function createDNSManagedZoneEntity(data: dns_v1.Schema$ManagedZone) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: DNS_MANAGED_ZONE_ENTITY_TYPE,
        _class: DNS_MANAGED_ZONE_ENTITY_CLASS,
        name: data.name,
        displayName: data.name as string,
        domainName: data.dnsName,
        visibility: data.visibility,
        kind: data.kind,
        // 3.3 Ensure that DNSSEC is enabled for Cloud DNS (Scored)
        dnssec: data.dnssecConfig?.state,
        // 3.4 Ensure that RSASHA1 is not used for the key-signing key in Cloud DNS DNSSEC (Not Scored)
        keySigningAlgorithm: data.dnssecConfig?.defaultKeySpecs?.find(
          (spec) => spec.keyType === 'keySigning',
        )?.algorithm,
        // 3.5 Ensure that RSASHA1 is not used for the zone-signing key in Cloud DNS DNSSEC (Not Scored)
        zoneSigningAlgorithm: data.dnssecConfig?.defaultKeySpecs?.find(
          (spec) => spec.keyType === 'zoneSigning',
        )?.algorithm,
        createdOn: parseTimePropertyValue(data.creationTime),
      },
    },
  });
}
