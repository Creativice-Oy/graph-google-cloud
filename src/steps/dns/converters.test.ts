import { getMockDNSManagedZone } from '../../../test/mocks';
import { createDNSManagedZoneEntity } from './converters';

describe('#createDNSManagedZoneEntity', () => {
  test('should convert to entity', () => {
    expect(
      createDNSManagedZoneEntity(getMockDNSManagedZone()),
    ).toMatchSnapshot();
  });

  test('should set dnssec to "on" when dnssec is enabled', () => {
    expect(
      createDNSManagedZoneEntity(
        getMockDNSManagedZone({
          dnssecConfig: {
            state: 'on',
            nonExistence: 'nsec3',
            kind: 'dns#managedZoneDnsSecConfig',
          },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('should set value to keySigningAlgorithm when it is present', () => {
    expect(
      createDNSManagedZoneEntity(
        getMockDNSManagedZone({
          dnssecConfig: {
            state: 'on',
            defaultKeySpecs: [
              {
                keyType: 'keySigning',
                algorithm: 'rsasha256',
                keyLength: 2048,
                kind: 'dns#dnsKeySpec',
              },
            ],
            nonExistence: 'nsec3',
            kind: 'dns#managedZoneDnsSecConfig',
          },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('should set value to zoneSigningAlgorithm when it is present', () => {
    expect(
      createDNSManagedZoneEntity(
        getMockDNSManagedZone({
          dnssecConfig: {
            state: 'on',
            defaultKeySpecs: [
              {
                keyType: 'zoneSigning',
                algorithm: 'rsasha256',
                keyLength: 1024,
                kind: 'dns#dnsKeySpec',
              },
            ],
            nonExistence: 'nsec3',
            kind: 'dns#managedZoneDnsSecConfig',
          },
        }),
      ),
    ).toMatchSnapshot();
  });
});
