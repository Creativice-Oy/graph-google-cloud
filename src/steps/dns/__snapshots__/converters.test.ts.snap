// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createDNSManagedZoneEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "DomainZone",
  ],
  "_key": "778386319953398157",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "creationTime": "2021-01-22T16:48:27.272Z",
        "description": "",
        "dnsName": "example.com.",
        "id": "778386319953398157",
        "kind": "dns#managedZone",
        "name": "example-zone",
        "nameServers": Array [
          "ns-cloud-a1.googledomains.com.",
          "ns-cloud-a2.googledomains.com.",
          "ns-cloud-a3.googledomains.com.",
          "ns-cloud-a4.googledomains.com.",
        ],
        "visibility": "public",
      },
    },
  ],
  "_type": "google_dns_managed_zone",
  "createdOn": 1611334107272,
  "description": "",
  "displayName": "example-zone",
  "dnssec": undefined,
  "domainName": "example.com.",
  "id": "778386319953398157",
  "keySigningAlgorithm": undefined,
  "kind": "dns#managedZone",
  "name": "example-zone",
  "visibility": "public",
  "zoneSigningAlgorithm": undefined,
}
`;

exports[`#createDNSManagedZoneEntity should set dnssec to "on" when dnssec is enabled 1`] = `
Object {
  "_class": Array [
    "DomainZone",
  ],
  "_key": "778386319953398157",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "creationTime": "2021-01-22T16:48:27.272Z",
        "description": "",
        "dnsName": "example.com.",
        "dnssecConfig": Object {
          "kind": "dns#managedZoneDnsSecConfig",
          "nonExistence": "nsec3",
          "state": "on",
        },
        "id": "778386319953398157",
        "kind": "dns#managedZone",
        "name": "example-zone",
        "nameServers": Array [
          "ns-cloud-a1.googledomains.com.",
          "ns-cloud-a2.googledomains.com.",
          "ns-cloud-a3.googledomains.com.",
          "ns-cloud-a4.googledomains.com.",
        ],
        "visibility": "public",
      },
    },
  ],
  "_type": "google_dns_managed_zone",
  "createdOn": 1611334107272,
  "description": "",
  "displayName": "example-zone",
  "dnssec": "on",
  "domainName": "example.com.",
  "id": "778386319953398157",
  "keySigningAlgorithm": undefined,
  "kind": "dns#managedZone",
  "name": "example-zone",
  "visibility": "public",
  "zoneSigningAlgorithm": undefined,
}
`;

exports[`#createDNSManagedZoneEntity should set value to keySigningAlgorithm when it is present 1`] = `
Object {
  "_class": Array [
    "DomainZone",
  ],
  "_key": "778386319953398157",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "creationTime": "2021-01-22T16:48:27.272Z",
        "description": "",
        "dnsName": "example.com.",
        "dnssecConfig": Object {
          "defaultKeySpecs": Array [
            Object {
              "algorithm": "rsasha256",
              "keyLength": 2048,
              "keyType": "keySigning",
              "kind": "dns#dnsKeySpec",
            },
          ],
          "kind": "dns#managedZoneDnsSecConfig",
          "nonExistence": "nsec3",
          "state": "on",
        },
        "id": "778386319953398157",
        "kind": "dns#managedZone",
        "name": "example-zone",
        "nameServers": Array [
          "ns-cloud-a1.googledomains.com.",
          "ns-cloud-a2.googledomains.com.",
          "ns-cloud-a3.googledomains.com.",
          "ns-cloud-a4.googledomains.com.",
        ],
        "visibility": "public",
      },
    },
  ],
  "_type": "google_dns_managed_zone",
  "createdOn": 1611334107272,
  "description": "",
  "displayName": "example-zone",
  "dnssec": "on",
  "domainName": "example.com.",
  "id": "778386319953398157",
  "keySigningAlgorithm": "rsasha256",
  "kind": "dns#managedZone",
  "name": "example-zone",
  "visibility": "public",
  "zoneSigningAlgorithm": undefined,
}
`;

exports[`#createDNSManagedZoneEntity should set value to zoneSigningAlgorithm when it is present 1`] = `
Object {
  "_class": Array [
    "DomainZone",
  ],
  "_key": "778386319953398157",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "creationTime": "2021-01-22T16:48:27.272Z",
        "description": "",
        "dnsName": "example.com.",
        "dnssecConfig": Object {
          "defaultKeySpecs": Array [
            Object {
              "algorithm": "rsasha256",
              "keyLength": 1024,
              "keyType": "zoneSigning",
              "kind": "dns#dnsKeySpec",
            },
          ],
          "kind": "dns#managedZoneDnsSecConfig",
          "nonExistence": "nsec3",
          "state": "on",
        },
        "id": "778386319953398157",
        "kind": "dns#managedZone",
        "name": "example-zone",
        "nameServers": Array [
          "ns-cloud-a1.googledomains.com.",
          "ns-cloud-a2.googledomains.com.",
          "ns-cloud-a3.googledomains.com.",
          "ns-cloud-a4.googledomains.com.",
        ],
        "visibility": "public",
      },
    },
  ],
  "_type": "google_dns_managed_zone",
  "createdOn": 1611334107272,
  "description": "",
  "displayName": "example-zone",
  "dnssec": "on",
  "domainName": "example.com.",
  "id": "778386319953398157",
  "keySigningAlgorithm": undefined,
  "kind": "dns#managedZone",
  "name": "example-zone",
  "visibility": "public",
  "zoneSigningAlgorithm": "rsasha256",
}
`;
